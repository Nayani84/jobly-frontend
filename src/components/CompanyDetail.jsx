import React, { useEffect, useState, useContext } from "react";
import { useParams } from "react-router-dom";
import JoblyApi from "../JoblyApi";
import JobCard from "./JobCard";
import { CurrentUserContext } from "../CurrentUserContext";
import './CompanyDetail.css';


function CompanyDetail({ applyToJob }) {
  const { handle } = useParams();
  const [company, setCompany] = useState(null);
  const { currentUser } = useContext(CurrentUserContext);

  // Fetch the company details on mount
  useEffect(() => {
    async function fetchCompany() {
      try {
        const company = await JoblyApi.getCompany(handle);
        setCompany(company);
      } catch (err) {
        console.error("Error fetching company data:", err);
      }
    }
    fetchCompany();
  }, [handle]);

  if (!company) return <p>Loading...</p>;

  return (
    <div className="CompanyDetail">
      <h2>{company.name}</h2>
      <p className="ComSum">{company.description}</p>
      <ul>
        {company.jobs.map(job => (
          <JobCard
            key={job.id}
            job={job}
            applyToJob={applyToJob}
            hasApplied={currentUser?.applications?.has?.(job.id) || false} />
        ))}
      </ul>
    </div>
  );
}

export default CompanyDetail;
