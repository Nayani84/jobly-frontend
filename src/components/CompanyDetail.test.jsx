import React from "react";
import { render, screen, waitFor } from "@testing-library/react";
import { MemoryRouter, Route, Routes } from "react-router-dom";
import CompanyDetail from "./CompanyDetail";
import { CurrentUserContext } from "../CurrentUserContext";
import JoblyApi from "../JoblyApi";
import { vi } from "vitest";

// Mock JoblyApi to simulate API responses
vi.mock("../JoblyApi");

const testCompany = {
  handle: "test-company",
  name: "Test Company",
  description: "This is a test company.",
  jobs: [
    { id: 1, title: "Software Engineer", salary: 100000, equity: "0.1" },
    { id: 2, title: "Data Scientist", salary: 120000, equity: "0.05" },
  ],
};

describe("CompanyDetail Component", () => {
  it("renders without crashing", async () => {
    JoblyApi.getCompany.mockResolvedValue(testCompany);

    render(
      <MemoryRouter initialEntries={["/companies/test-company"]}>
        <Routes>
          <Route
            path="/companies/:handle"
            element={
              <CurrentUserContext.Provider value={{ currentUser: { applications: new Set() } }}>
                <CompanyDetail applyToJob={vi.fn()} />
              </CurrentUserContext.Provider>
            }
          />
        </Routes>
      </MemoryRouter>
    );

    expect(await screen.findByText("Test Company")).toBeInTheDocument();
  });

  it("matches snapshot", async () => {
    JoblyApi.getCompany.mockResolvedValue(testCompany);

    const { asFragment } = render(
      <MemoryRouter initialEntries={["/companies/test-company"]}>
        <Routes>
          <Route
            path="/companies/:handle"
            element={
              <CurrentUserContext.Provider value={{ currentUser: { applications: new Set() } }}>
                <CompanyDetail applyToJob={vi.fn()} />
              </CurrentUserContext.Provider>
            }
          />
        </Routes>
      </MemoryRouter>
    );

    await waitFor(() => screen.getByText("Test Company"));
    expect(asFragment()).toMatchSnapshot();
  });

  it("displays company and job details correctly", async () => {
    JoblyApi.getCompany.mockResolvedValue(testCompany);

    render(
      <MemoryRouter initialEntries={["/companies/test-company"]}>
        <Routes>
          <Route
            path="/companies/:handle"
            element={
              <CurrentUserContext.Provider value={{ currentUser: { applications: new Set() } }}>
                <CompanyDetail applyToJob={vi.fn()} />
              </CurrentUserContext.Provider>
            }
          />
        </Routes>
      </MemoryRouter>
    );

    expect(await screen.findByText("Test Company")).toBeInTheDocument();
    expect(screen.getByText("This is a test company.")).toBeInTheDocument();
    expect(screen.getByText("Software Engineer")).toBeInTheDocument();
    expect(screen.getByText("Data Scientist")).toBeInTheDocument();
  });

  it("handles loading state", () => {
    render(
      <MemoryRouter initialEntries={["/companies/test-company"]}>
        <Routes>
          <Route
            path="/companies/:handle"
            element={
              <CurrentUserContext.Provider value={{ currentUser: { applications: new Set() } }}>
                <CompanyDetail applyToJob={vi.fn()} />
              </CurrentUserContext.Provider>
            }
          />
        </Routes>
      </MemoryRouter>
    );

    expect(screen.getByText("Loading...")).toBeInTheDocument();
  });
});
