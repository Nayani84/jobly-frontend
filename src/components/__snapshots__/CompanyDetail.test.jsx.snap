// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CompanyDetail Component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="CompanyDetail"
  >
    <h2>
      Test Company
    </h2>
    <p
      class="ComSum"
    >
      This is a test company.
    </p>
    <ul>
      <div
        class="JobCard"
      >
        <div>
          <h3>
            Software Engineer
          </h3>
          <p>
            Salary: $100000
          </p>
          <p>
            Equity: 0.1
          </p>
        </div>
        <button
          class="btn-apply"
        >
          Apply
        </button>
      </div>
      <div
        class="JobCard"
      >
        <div>
          <h3>
            Data Scientist
          </h3>
          <p>
            Salary: $120000
          </p>
          <p>
            Equity: 0.05
          </p>
        </div>
        <button
          class="btn-apply"
        >
          Apply
        </button>
      </div>
    </ul>
  </div>
</DocumentFragment>
`;
